### PR TITLE
HDDS-11391. Frequent Ozone DN Crashes During OM + DN Decommission with Freon

### PR DESCRIPTION
**Pull request by jojochuang**
## What changes were proposed in this pull request?
HDDS-11391. Frequent Ozone DN Crashes During OM + DN Decommission with Freon

Please describe your PR in detail:
* When in piggyback mode, check if WriteChunk/PutBlock has end of block field toggled. If so, close the corresponding block file open descriptor.
* Added unit test and integration test

TODO:
There is still one corner case not fixed yet: if client does write - hsync and then close (meaning write buffer is emptied), for some reason client doesn't toggle end of block field.
Unfortunately our BlockOutputStream implementation is becoming too complex to understand. Refactoring would be required later.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-11391
## How was this patch tested?

Unit tests, integration tests

### CODE REVIEW

File: hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/keyvalue/impl/AbstractTestChunkManager.java [hunk 1]
```diff
@@ -128,6 +135,47 @@ protected void checkChunkFileCount(int expected) {
     assertEquals(expected, files.length);
   }
 
+  /**
+   * Helper class to check if a file is in use.
+   */
+  public static class FuserCheck {
```
Comment (LZD-PratyushBhatt) [inline hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/keyvalue/impl/AbstractTestChunkManager.java:141]: Hi Wei, whats the reason behind using this as a class name?
FileUsageChecker or FileInUseChecker sounds more apt to me

File: hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/keyvalue/impl/AbstractTestChunkManager.java [hunk 2]
```diff
@@ -128,6 +135,47 @@ protected void checkChunkFileCount(int expected) {
...
+    public static boolean isFileNotInUse(String filePath) {
+      try {
+        Process process = new ProcessBuilder("fuser", filePath).start();
+        BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()));
+        return reader.readLine() == null;  // If fuser returns no output, the file is not in use
+      } catch (IOException e) {
+        LOG.warn("Failed to check if file is in use: {}", filePath, e);
+        return false;  // On failure, assume the file is in use
+      }
```
Comment (LZD-PratyushBhatt) [inline hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/keyvalue/impl/AbstractTestChunkManager.java:150]: Should we add a finally block to cleanup, but we would need to define process outside the try block then:
`finally {
            if (process != null) {
                process.destroy();
            }
        }`

File: hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/keyvalue/impl/AbstractTestChunkManager.java (hunk 1 above)
Comment (jojochuang) [reply hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/keyvalue/impl/AbstractTestChunkManager.java]: removed the static class. Just use the method itself.

**Review by ashishkumar50 — COMMENTED** [review]
Thanks @jojochuang for the fix, change LGTM +1.

**Comment by jojochuang** [comment]
Merge based on Ashish's +1.

**Pull request merged**